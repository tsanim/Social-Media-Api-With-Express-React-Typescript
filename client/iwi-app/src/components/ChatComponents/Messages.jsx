import React from 'react';
import ScrollToBottom from 'react-scroll-to-bottom';
import PropTypes from 'prop-types';
import Message from './Message';

function Messages(props) {
    return (
        <ScrollToBottom className="messagesContainer">
            <p className="msg">{props.infoMessage}</p>
            <ul>
                {
                    props.messages.map((m, i) => {
                        let liClass = i === 0 ? 'clearfix' : null;
                        let isMine = m.creator._id === localStorage.getItem('userId');

                        return (
                            (<li key={i} className={liClass}>
                                <Message 
                                    isMine={isMine}
                                    message={m}
                                />
                            </li>)
                        )
                    })
                }
            </ul>
        </ScrollToBottom>
    )
}

Messages.propTypes = {
    infoMessage: PropTypes.string,
    messages: PropTypes.array,
}

export default Messages;