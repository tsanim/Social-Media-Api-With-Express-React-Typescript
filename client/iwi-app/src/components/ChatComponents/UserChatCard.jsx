import React from 'react';
import URI from '../../config/config';
import PropTypes from 'prop-types';

function UserChatCard(props) {
    return (
        <figure className={ props.isMine ? "float-right" : null }>
            <img src={`${URI}/feed/image/${props.onlineUser.imageId}`} alt="userImg" />
            <div className="names">
                <figcaption>{props.onlineUser.username}</figcaption>
            </div>
        </figure>
    )
}

UserChatCard.propTypes = {
    onlineUser: PropTypes.object,
    isMine: PropTypes.bool
}


export default UserChatCard;