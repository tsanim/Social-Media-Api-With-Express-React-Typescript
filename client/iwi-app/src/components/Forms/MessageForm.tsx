import React from 'react';
import PropTypes from 'prop-types';
import MessageFormProps from '../../interfaces/Components/Forms/MessageFormProps.interface';

function MessageForm(props: MessageFormProps) {
    return (
        <div className="messageForm">
            <form id="messageForm">
                <input
                    type="text"
                    placeholder="Type a message"
                    name="messageText"
                    value={props.messageText}
                    onChange={props.onInputChangeHandler}
                    onKeyPress={props.onKeyPressHandler}
                    onKeyDown={props.onKeyDownHandler}
                    onBlur={props.onBlurHandler}
                />
                <input type="button" onClick={props.sendMessageHandler} value="Send" />
            </form>
        </div>
    )
}

MessageForm.propTypes = {
    messageText: PropTypes.string,
    onInputChangeHandler: PropTypes.func,
    onKeyPressHandler: PropTypes.func,
    onFocusHandler: PropTypes.func,
    onBlurHandler: PropTypes.func,
    sendMessageHandler: PropTypes.func,
    onKeyDownHandler: PropTypes.func,
}

export default MessageForm;
