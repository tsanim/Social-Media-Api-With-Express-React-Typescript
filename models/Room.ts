import mongoose, { Schema } from 'mongoose';
import IRoom from '../interfaces/models/Room.interface';

const roomSchema: Schema = new Schema({
    pairUsers: [{ type: Schema.Types.ObjectId, ref: 'User' }],
    messages: [{ type: Schema.Types.ObjectId, ref: 'Message' }],
});

const Room = mongoose.model<IRoom>('Room', roomSchema);

export default Room;