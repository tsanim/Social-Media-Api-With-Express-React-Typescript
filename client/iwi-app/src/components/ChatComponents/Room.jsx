import React, { useEffect } from 'react';
import Messages from './Messages';
import MessageForm from '../Forms/MessageForm';
import PropTypes from 'prop-types';
import UserChatCard from './UserChatCard';

function Room(props) {
    useEffect(() => {
        return () => {
            props.onUnmountHandler();
        }
    }, []);

    return (
        <div className="chat">
            <header className="chatHeader">
                <UserChatCard 
                    onlineUser={props.onlineUser}
                />
            </header>
            <Messages
                onlineUser={props.onlineUser}
                messages={props.messages}
                infoMessage={props.infoMessage}
            />
            <span className="typing">{props.typingMessage}</span>
            <MessageForm 
                messageText={props.messageText}
                onInputChangeHandler={props.onInputChangeHandler}
                onBlurHandler={props.onBlurHandler}
                onKeyDownHandler={props.onKeyDownHandler}
                onKeyPressHandler={props.onKeyPressHandler}
                sendMessageHandler={props.sendMessageHandler}
            />
        </div>
    )
}

Room.propTypes = {
    onlineUser: PropTypes.object,
    messages: PropTypes.array,
    infoMessage: PropTypes.string,
    typingMessage: PropTypes.string,
    onInputChangeHandler: PropTypes.func,
    onKeyPressHandler: PropTypes.func,
    onKeyDownHandler: PropTypes.func,
    onBlurHandler: PropTypes.func,
    messageText: PropTypes.string,
    sendMessageHandler: PropTypes.func,
    onUnmountHandler: PropTypes.func
}


export default Room;