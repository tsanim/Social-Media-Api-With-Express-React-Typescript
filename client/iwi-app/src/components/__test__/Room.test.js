import React from 'react';
import Room from '../ChatComponents/Room';

import { shallow } from 'enzyme';
import Messages from '../ChatComponents/Messages';
import MessageForm from '../Forms/MessageForm';
import UserChatCard from '../ChatComponents/UserChatCard';

function roomComponentSetup() {
    const props = {
        onlineUser: {},
        messages: [],
        infoMessage: '',
        typing: '',
        onInputChangeHandler: jest.fn(),
        onKeyPressHandler: jest.fn(),
        onKeyDownHandler: jest.fn(),
        onBlurHandler: jest.fn(),
        message: '',
        sendMessageHandler: jest.fn(),
        onUnmountHandler: jest.fn()
    }

    const enzymeWrapper = shallow(<Room {...props} />)

    return {
        props,
        enzymeWrapper
    }
}

describe('Room component', () => {
    it('should render initial layout', () => {
        const { enzymeWrapper } = roomComponentSetup();

        expect(enzymeWrapper.getElements()).toMatchSnapshot();
    });

    it('should render container div tag with class "chat"', () => {
        const { enzymeWrapper } = roomComponentSetup();

        expect(enzymeWrapper.find('div.chat')).toHaveLength(1);
    });

    it('should render header tag with class "chatHeader"', () => {
        const { enzymeWrapper } = roomComponentSetup();

        expect(enzymeWrapper.find('header.chatHeader')).toHaveLength(1);
    });

    it('should render Messages Component', () => {
        const { enzymeWrapper } = roomComponentSetup();

        expect(enzymeWrapper.find(Messages)).toHaveLength(1);
    });

    it('should render Message Component', () => {
        const { enzymeWrapper } = roomComponentSetup();

        expect(enzymeWrapper.find(MessageForm)).toHaveLength(1);
    });

    it('should render span tag with class "typing"', () => {
        const { enzymeWrapper } = roomComponentSetup();

        expect(enzymeWrapper.find('span.typing')).toHaveLength(1);
    });

    it('should render UserChatCard Component', () => {
        const { enzymeWrapper } = roomComponentSetup();

        expect(enzymeWrapper.find(UserChatCard)).toHaveLength(1);
    });
})