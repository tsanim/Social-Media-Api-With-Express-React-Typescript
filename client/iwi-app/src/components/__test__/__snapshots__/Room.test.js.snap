// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Room component should render initial layout 1`] = `
Array [
  <div
    className="chat"
  >
    <header
      className="chatHeader"
    >
      <UserChatCard
        onlineUser={Object {}}
      />
    </header>
    <Messages
      infoMessage=""
      messages={Array []}
      onlineUser={Object {}}
    />
    <span
      className="typing"
    />
    <MessageForm
      onBlurHandler={[MockFunction]}
      onInputChangeHandler={[MockFunction]}
      onKeyDownHandler={[MockFunction]}
      onKeyPressHandler={[MockFunction]}
      sendMessageHandler={[MockFunction]}
    />
  </div>,
]
`;
